cpu/atmega_common/periph/rtc.c

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright (C) 2018 Acutam Automation, LLC
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_atmega_common
+ * @ingroup     drivers_periph_rtc
+ *
+ * @{
+ *
+ * @file
+ * @brief       RTC interface wrapper for use with RTT modules.
+ *
+ * @author      Matthew Blue <matthew.blue.neuro@gmail.com>
+ *
+ * In order to safely sleep when using the RTT:
+ * 1. Disable interrupts
+ * 2. Write to one of the asynch registers (e.g. TCCR2A)
+ * 3. Wait for ASSR register's busy flags to clear
+ * 4. Re-enable interrupts
+ * 5. Sleep before interrupt re-enable takes effect
+ *
+ * @}
+ */
+
+#include <avr/interrupt.h>
+#include <time.h>
+
+#include "cpu.h"
+#include "irq.h"
+#include "periph/rtc.h"
+#include "periph/rtt.h"
+#include "periph_conf.h"
+#include "thread.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+/* guard file in case no RTC device is defined */
+#if RTC_NUMOF
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct {
+    time_t alarm;               /* alarm_cb when time == alarm */
+    rtc_alarm_cb_t alarm_cb;    /* callback called from RTC alarm */
+    void *alarm_arg;            /* argument passed to the callback */
+} rtc_state_t;
+
+static rtc_state_t rtc_state;
+
+/* RTC second counter. This is treated as-if it is a register as it is
+ * incremented by the overflow ISR. This MUST BE marked volatile for two
+ * reasons: to guarantee ordering with other register operations, and to
+ * force redundant access to occur (in case it was changed by the ISR). */
+static volatile time_t time_cnt;
+
+static inline void _asynch_wait(void)
+{
+    /* Wait until all busy flags clear. According to the datasheet,
+     * this can take up to 2 positive edges of TOSC1 (32kHz). */
+    while (ASSR & ((1 << TCN2UB) | (1 << OCR2AUB) | (1 << OCR2BUB)
+                   | (1 << TCR2AUB) | (1 << TCR2BUB)));
+}
+
+/* This safely reads time_cnt and TCNT2, since an atomic read cannot be provided.
+ * Note: still requires prior synch using _asynch_wait() */
+static inline time_t _safe_time_get(void)
+{
+    time_t time_tmp;
+    uint8_t irq_flag, cnt_tmp;
+
+    do {
+        /* Operations must occur in this order: */
+        irq_flag = TIFR2 & (1 << TOV2);
+        time_tmp = time_cnt;
+        cnt_tmp = TCNT2;
+
+    /* Detect rollover after time_tmp read. If condition is met, then rollover
+     * just occured, and it is safe to read again now. */
+    } while ((time_tmp != time_cnt) || (irq_flag != (TIFR2 & (1 << TOV2))));
```

Comment:
I´m quite sure that `TOV2` will ever be 0. But i might be wrong.

```
Bit 0 – TOV: Timer/Counter2, Overflow Flag
The TOV bit is set (one) when an overflow occurs in Timer/Counter2. TOV is cleared by hardware when
executing the corresponding interrupt handling vector.
```
As the interrupt is enabled `TOV2` will be cleared. Polling it only is usefull when no `IRQ routine` is used, imho. 
Maybe handling like in xtimer would be an idea.

https://github.com/RIOT-OS/RIOT/blob/8755a26716009958148f9b59440d2f6f3f75b0bc/sys/xtimer/xtimer_core.c#L89

Reply:
`TOV2` is checked in case interrupts are disabled. These two checks are not really polls, because the condition is only expected to occur zero or one times. It is just more space efficient than using `if`, since the compiler is not allowed to optimize this due to volatile.